Neo4j Hybrid Search

vga91 opened the pull request:
<!--
Thank you so much for your contribution!

Please fill in all the sections below.
Please open the PR as a draft initially. Once it is reviewed and approved, we will ask you to add documentation and examples.
Please note that PRs with breaking changes or without tests will be rejected.

Please note that PRs will be reviewed based on the priority of the issues they address.
We ask for your patience. We are doing our best to review your PR as quickly as possible.
Please refrain from pinging and asking when it will be reviewed. Thank you for understanding!
-->

## Issue
<!-- Please specify the ID of the issue this PR is addressing. For example: "Closes #1234" or "Fixes #1234" -->
Closes `Add Hybrid Search with re-ranker, fulltext index generation` checklist's item of #111

Closes https://github.com/langchain4j/langchain4j/issues/1306

## Change
<!-- Please describe the changes you made. -->

Migration from langchain4j core repo's PR: https://github.com/langchain4j/langchain4j/pull/2575

- If we have [these configs](https://github.com/langchain4j/langchain4j-community/pull/113/files#diff-1b0372976e6deaaff22a9749ae82a8f561684bdb244ff371b60574353283bb1aR21) we can execute an hybrid search with both vector and fullText index

- Created a test replicating as far as possible [the provided langchaing-ai example](https://neo4j.com/developer-blog/enhance-rag-knowledge-graph/).
  - Difficult something more meaningful since several things are missing compared to langchain python::
  - `Neo4jVector.from_existing_graph`
  - `TokenTextSplitter`
  - `ChatPromptTemplate.from_template`
  - `llm_transformer.convert_to_graph_documents`


- Created tests with different configs and invalid ones

#### TO EVALUATE

- Maybe we can use ChatMemoryStore to store the FullTextIndexes
- The issue says `WITH collect({node:node, score:score})` to retrieve both vector.queryNodes and fullText.queryNodes results, 
but we currently have `YIELD node, score WHERE score >= $minScore` to retrieve  vector.queryNodes  instead. 
So, used this method also for fullText.
Alternatively, we can change vector.queryNodes as well, but it would probably have to be done in another separate issue, where we should test `max` result as well, [as done here](https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/vectorstores/neo4j_vector.py#L115)



## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] There are no breaking changes
- [ ] I have added unit and integration tests for my change
- [ ] I have manually run all the unit tests in _all_ modules, and they are all green
- [ ] I have manually run all integration tests in the module I have added/changed, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-community-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

Martin7-1 commented on embedding-stores/langchain4j-community-neo4j/src/main/java/dev/langchain4j/community/store/embedding/neo4j/Builder.java:
```diff
@@ -18,6 +18,10 @@ public class Builder {
+    private String fullTextIndexName;
+    private String fullTextQuery;
+    private String fullTextRetrievalQuery;
+    private boolean fullTextAutocreate;
```
```suggestion
    private boolean autoCreateFullText;
```

WDYT?

Martin7-1 commented on embedding-stores/langchain4j-community-neo4j/src/main/java/dev/langchain4j/community/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -1,17 +1,6 @@
-import static dev.langchain4j.community.store.embedding.neo4j.Neo4jEmbeddingUtils.getRowsBatched;
-import static dev.langchain4j.community.store.embedding.neo4j.Neo4jEmbeddingUtils.sanitizeOrThrows;
-import static dev.langchain4j.community.store.embedding.neo4j.Neo4jEmbeddingUtils.toEmbeddingMatch;
+import static dev.langchain4j.community.store.embedding.neo4j.Neo4jEmbeddingUtils.*;
```
Let's just keep all imports explicit.

Martin7-1 commented on embedding-stores/langchain4j-community-neo4j/src/main/java/dev/langchain4j/community/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -101,6 +93,11 @@ public class Neo4jEmbeddingStore implements EmbeddingStore<TextSegment> {
+    private final String fullTextIndexName;
+    private final String fullTextQuery;
+    private final String fullTextRetrievalQuery;
+    private final boolean fullTextAutocreate;
```
Same as above.

vga91 replied:
changed

Martin7-1 commented on embedding-stores/langchain4j-community-neo4j/src/main/java/dev/langchain4j/community/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -362,6 +391,21 @@ var record = resIndex.single();
 
+    private void createFullTextIndex() {
+        if (!fullTextAutocreate) {
+            System.out.println("Neo4jEmbeddingStore.createFullTextIndex");
```
Is it a debug-statement?

vga91 replied:
sorry didn't see it, removed

Martin7-1 commented on embedding-stores/langchain4j-community-neo4j/src/main/java/dev/langchain4j/community/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -362,6 +391,21 @@ var record = resIndex.single();
+        try (var session = session()) {
+
+            final String query = "CREATE FULLTEXT INDEX %s IF NOT EXISTS FOR (n:%s) ON EACH [n.%s]"
+                    .formatted(this.fullTextIndexName, this.sanitizedLabel, this.sanitizedIdProperty);
```
`.formatted()` will cause Android's compatible problem, see https://github.com/langchain4j/langchain4j/pull/2355

vga91 replied:
changed

Martin7-1 commented on embedding-stores/langchain4j-community-neo4j/src/main/java/dev/langchain4j/community/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -258,25 +269,42 @@ public EmbeddingSearchResult<TextSegment> search(EmbeddingSearchRequest request)
+
+            if (fullTextQuery != null) {
+
+                query +=
```
I'm not familiar with Neo4j, it is possible to encounter problem like SQL inject when using string concatencation?

vga91 replied:
Good point, I added a query validation (by checking the column names) and a `session.executeRead` instead of `session.run`,
so that if the query is not read-only it will fail

Martin7-1 commented on embedding-stores/langchain4j-community-neo4j/src/main/java/dev/langchain4j/community/store/embedding/neo4j/Neo4jEmbeddingStore.java:
```diff
@@ -258,25 +269,42 @@ public EmbeddingSearchResult<TextSegment> search(EmbeddingSearchRequest request)
 
         try (var session = session()) {
-            Map<String, Object> params = Map.of(
+            Map<String, Object> params = new HashMap<>(Map.of(
```
nit: Using immutable `Map.of` is better, any other intention to use `new HashMap<>()`?

vga91 replied:
I added the `new HashMap<>` cause there is a `params.putAll` below.
I can change it to something else if necessary?

Martin7-1 replied:
Oh. I didn't see it. Just keep it :)

Martin7-1 reviewed (commented):
@vga91 Thank you!

Martin7-1 commented on embedding-stores/langchain4j-community-neo4j/pom.xml:
```diff
@@ -87,5 +87,35 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j</artifactId>
+            <version>1.0.0-alpha2-SNAPSHOT</version>
```
```suggestion
            <version>${project.version}</version>
```

Martin7-1 commented on embedding-stores/langchain4j-community-neo4j/pom.xml:
```diff
@@ -87,5 +87,35 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-document-loader-selenium</artifactId>
+            <version>1.0.0-beta3-SNAPSHOT</version>
```
```suggestion
            <version>${project.version}</version>
```

Martin7-1 commented on embedding-stores/langchain4j-community-neo4j/pom.xml:
```diff
@@ -87,5 +87,35 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-document-transformer-jsoup</artifactId>
+            <version>1.0.0-beta3-SNAPSHOT</version>
```
```suggestion
            <version>${project.version}</version>
```

Martin7-1 commented on embedding-stores/langchain4j-community-neo4j/pom.xml:
```diff
@@ -87,5 +87,35 @@
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-open-ai</artifactId>
+            <version>1.0.0-beta3-SNAPSHOT</version>
```
```suggestion
            <version>${project.version}</version>
```

Martin7-1 reviewed (commented):
@vga91 Thank you! BTW, could you please move `Builder` into `Neo4jEmbeddingStore` and make it a static inner class (to keep the same style as the other modules)?

vga91 commented:
> @vga91 Thank you! BTW, could you please move `Builder` into `Neo4jEmbeddingStore` and make it a static inner class (to keep the same style as the other modules)?

Moved the `Builder` and changed the `${project.version}`s. Thanks!

Martin7-1 commented:
The `should_emulate_issue_1306_case` throws:

```
org.neo4j.driver.exceptions.ClientException: Invalid input 'I': expected a graph pattern, a parameter, ')', ':', 'IS', 'WHERE' or '{' (line 1, column 19 (offset: 18))
"CREATE (Elizabeth I:Entity {id: "Elizabeth I"}),"
                   ^

	at org.neo4j.driver.internal.util.Futures.blockingGet(Futures.java:108)
	at org.neo4j.driver.internal.InternalTransaction.run(InternalTransaction.java:55)
	at org.neo4j.driver.internal.AbstractQueryRunner.run(AbstractQueryRunner.java:32)
	at org.neo4j.driver.internal.AbstractQueryRunner.run(AbstractQueryRunner.java:47)
	at org.neo4j.driver.internal.DelegatingTransactionContext.run(DelegatingTransactionContext.java:51)
	at dev.langchain4j.community.store.embedding.neo4j.Neo4jEmbeddingStoreIT.lambda$should_emulate_issue_1306_case$2(Neo4jEmbeddingStoreIT.java:112)
	at org.neo4j.driver.internal.InternalSession.lambda$execute$4(InternalSession.java:161)
	at org.neo4j.driver.internal.InternalSession.lambda$transaction$5(InternalSession.java:178)
	at org.neo4j.driver.internal.retry.ExponentialBackoffRetryLogic.retry(ExponentialBackoffRetryLogic.java:113)
	at org.neo4j.driver.internal.InternalSession.transaction(InternalSession.java:175)
	at org.neo4j.driver.internal.InternalSession.execute(InternalSession.java:160)
	at org.neo4j.driver.internal.InternalSession.executeWrite(InternalSession.java:132)
	at org.neo4j.driver.Session.executeWrite(Session.java:169)
	at dev.langchain4j.community.store.embedding.neo4j.Neo4jEmbeddingStoreIT.should_emulate_issue_1306_case(Neo4jEmbeddingStoreIT.java:111)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	Suppressed: org.neo4j.driver.internal.util.ErrorUtil$InternalExceptionCause
		at org.neo4j.driver.internal.util.ErrorUtil.newNeo4jError(ErrorUtil.java:76)
		at org.neo4j.driver.internal.async.inbound.InboundMessageDispatcher.handleFailureMessage(InboundMessageDispatcher.java:111)
		at org.neo4j.driver.internal.messaging.common.CommonMessageReader.unpackFailureMessage(CommonMessageReader.java:62)
		at org.neo4j.driver.internal.messaging.common.CommonMessageReader.read(CommonMessageReader.java:46)
		at org.neo4j.driver.internal.async.inbound.InboundMessageHandler.channelRead0(InboundMessageHandler.java:78)
		at org.neo4j.driver.internal.async.inbound.InboundMessageHandler.channelRead0(InboundMessageHandler.java:34)
		at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
		at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
		at org.neo4j.driver.internal.async.inbound.MessageDecoder.channelRead(MessageDecoder.java:40)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
		at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
		at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:333)
		at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:455)
		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
		at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:289)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796)
		at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:732)
		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:658)
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
		at java.base/java.lang.Thread.run(Thread.java:840)
```

Could you please check?

vga91 commented:
> The `should_emulate_issue_1306_case` throws:

> org.neo4j.driver.exceptions.ClientException: Invalid input 'I': expected a graph pattern, a parameter, ')', ':', 'IS', 'WHERE' or '{' (line 1, column 19 (offset: 18))
> "CREATE (Elizabeth I:Entity {id: "Elizabeth I"}),"
>                    ^
> 

> ```
> 
> Could you please check?

I fixed the tests, now it is more deterministic.
Also, I removed the `SchemaNames.sanitize` from the `this.fullTextQuery` assign. 
Besides being unnecessary, it is incorrect since it can wrap it with backtick, e.g. "\`elizabeth*\`" instead of "elizabeth*"

And also, improved the [null embedding handling](https://github.com/langchain4j/langchain4j-community/pull/113/commits/71c44dc609cd2d02bc765ce6cb6926c85319eb75#diff-a73aed9cb0a7981cb18cf64fce5095488d5db89eafa914d023ec99ab5260ce71R52), otherwise we could have a
```
org.neo4j.driver.exceptions.value.Uncoercible: Cannot coerce NULL to Java List
```

Martin7-1 reviewed (approved):
@vga91 Thank you!
